### ISSUE TITLE
FSharp.Compiler.Service dependency on MSBuild

### ISSUE THREAD

**Issue by nightroman** [fsharp/FSharp.Compiler.Service#631]
As far as I know, project related tools moved to FSharp.Compiler.Service.ProjectCracker. At the same time, FSharp.Compiler.Service still depends on MSBuild and cannot be loaded and used without MSBuild installed.

ILSpy shows the following references:

```
Microsoft.Build.Framework, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Microsoft.Build.Tasks.v12.0, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Microsoft.Build.Utilities.v12.0, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

I wonder if this dependency is by design or not. In the latter case, it would be nice if this dependency is removed. FSharp.Compiler.Service is useful on its own without MSBuild related stuff.

**Comment by tpetricek** [comment]
I was just going to report this too - Fable, which uses FCS 6.0.2 still does not work on machine with just VS 2015 (you need to install MSBUILD to be able to run it).

**Comment by tpetricek** [comment]
There was earlier discussion about this, but apparently, this is still an issue: https://github.com/fsharp/FSharp.Compiler.Service/issues/337

**Comment by 7sharp9** [comment]
Isn't msbuild used for reference resolution too, except in the case of mono which uses simple resolution.

**Comment by nosami** [comment]
Yeah it's used for reference resolution. Would really like to drop the
MSBuild deps though if possible. Ideally, the reference resolution should
be done before any FCS interactions.

**Comment by tpetricek** [comment]
It would be great to solve this - as more and more projects use FSC and not many people use VS 2013, this is a growing issue. If we don't solve this before Progressive F# tutorials in London, I bet many people coming to Suave and Fable tutorials will be affected by this (with Suave, FSC can provide nice live-reloading experience and Fable relies on this directly).

What is a minimal viable solution to make sure we do not give really crappy experience? Should I recover my PR to bundle the DLLs as part of the package?

**Comment by 7sharp9** [comment]
I wonder why its so hard to rip out, desperately need a #NO_MSBUILD flag to purge it so reference behavior is mono stylee.

**Comment by tpetricek** [comment]
(more discussion about the earlier PR is here https://github.com/fsharp/FSharp.Compiler.Service/pull/338 - yes, it's not nice solution, but I think we need _some_ solution)

**Comment by dsyme** [comment]
Let's just factor the dependency into another DLL (so you can opt in to Visual F#-Tools compatible MSBuild support).

**Comment by dsyme** [comment]
@nightroman  @tpetricek @7sharp9 @nosami  See https://github.com/fsharp/FSharp.Compiler.Service/pull/649, please take a look over this.

**Comment by nightroman** [comment]
@dsyme, it looks very promising, thank you very much! This approach, with some minor adjustments on my side, should work for [FSharpFar](https://github.com/nightroman/FarNet/tree/master/FSharpFar). This module supports two configuration types, .fsproj and .fs.ini. The latter is effectively just arguments of fsc and fsi in a friendly format with some helpers like environment variables expansion. As far as understand, .fs.ini scenario will be able to work without MSBuild. Developers of F# scripts for Far Manager may still use .fsproj with MSBuild / Visual Studio / VS Code. But F# scripts released for Far Manager users come with .fs.ini, so that users do not have to install MSBuild. This is exactly what I needed.

**Comment by tpetricek** [comment]
@dsyme Thank you for looking into this!! Solving this will be fantastic. I should be able to look once I'm done with the training I'm doing over the next two days!

**Comment by nightroman** [comment]
@tpetricek, it would be interesting to know if it works for you. So I far I cannot make it working and I am not sure if it is my fault or not. *...\Microsoft.Build.Framework\v4.0_14.0.0.0__b03f5f7f11d50a3a\Microsoft.Build.Framework.dll* is loaded even if I opt out, if it is present, and it fails if it is not.

**Comment by dsyme** [comment]
@nightroman Could you submit repro steps?  In particular, which tool is hosting FSHarp.Compiler.Service?

**Comment by nightroman** [comment]
[FSharpFar]: https://github.com/nightroman/FarNet/tree/master/FSharpFar

@dsyme The tool is [FSharpFar] and the FCS scenarios that it uses are briefly
described four comments above.

Let me put the latest changes of FSharpFar in order and release its version
that uses FCS 8.0.0 and opts out of MSBuild in cases of `.fs.ini` configuration.
Then I will describe how to get Far Manager, FarNet, and FSharpFar and repro steps.

**Comment by nightroman** [comment]
@dsyme 


I am trying to investigate, too. Here is an interesting result.

I added this brutal check and exception:

````fsharp
    let internal GetBestAvailableResolver(msbuildEnabled: bool) =
        if msbuildEnabled then invalidOp "unexpected msbuildEnabled"
````

Then I ran my case with `msbuildEnabled = false`.
I got the exception which I kind of expected. Here is the stack:

````
    TypeInitializationException:
    The type initializer for 'Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker' threw an exception.

    TypeInitializationException:
    The type initializer for '<StartupCode$FSharp-Compiler-Service>.$Service' threw an exception.

    InvalidOperationException:
    unexpected msbuildEnabled

    System.TypeInitializationException: The type initializer for 'Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker' threw an exception. ---> System.TypeInitializationException: The type initializer for '<StartupCode$FSharp-Compiler-Service>.$Service' threw an exception. ---> System.InvalidOperationException: unexpected msbuildEnabled
       at Microsoft.FSharp.Compiler.SimulatedMSBuildReferenceResolver.GetBestAvailableResolver(Boolean msbuildEnabled)
       at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker.Create(FSharpOption`1 projectCacheSize, FSharpOption`1 keepAssemblyContents, FSharpOption`1 keepAllBackgroundResolutions, FSharpOption`1 msbuildEnabled)
       at <StartupCode$FSharp-Compiler-Service>.$Service..cctor()
       --- End of inner exception stack trace ---
       at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker..cctor()
       --- End of inner exception stack trace ---
       at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker.Create(FSharpOption`1 projectCacheSize, FSharpOption`1 keepAssemblyContents, FSharpOption`1 keepAllBackgroundResolutions, FSharpOption`1 msbuildEnabled)
       at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession..ctor(FsiEvaluationSessionHostConfig fsi, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, Boolean fsiCollectible, Boolean msbuildEnabled)
       at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession.Create(FsiEvaluationSessionHostConfig fsiConfig, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, FSharpOption`1 collectible, FSharpOption`1 msbuildEnabled)
       at FSharpFar.Session.Session..ctor(String configFile)
       at FSharpFar.Session.Session.FindOrCreate(String path)
       at FSharpFar.FarCommand.Invoke(Object sender, ModuleCommandEventArgs e)
       at FarNet.Far0.InvokeCommand(Char* command, Boolean isMacro)
       at FarNet.Far0.AsOpen(OpenInfo* info)
       at OpenW(OpenInfo* info)
````

The static initializer

````fsharp
    static let globalInstance = FSharpChecker.Create()
````

calls `Create` with the default parameters, no matter what I actually provide in my other calls.

Before I go further, I would like to know if this call with default parameters is by design or not.

The above does not mean that it all is not working on a machine without MSBuild.
Why it fails on my machine without MSBuild is yet unknown, please ignore for now.

**Comment by nightroman** [comment]
I cannot tell exactly what is wrong on a machine without MSBuild. But according
to `procmon`, in spite of `msbuildEnabled = false`, *Microsoft.Build.Framework*
is searched in the registry and GAC.

It fails in try/with here

FSharp.Compiler.Service\src\fsharp\fsi\fsi.fs:

````fsharp
    /// The single, global interactive checker that can be safely used in conjunction with other operations
    /// on the FsiEvaluationSession. 
    let checker = FSharpChecker.Create()

    let (tcGlobals,frameworkTcImports,nonFrameworkResolutions,unresolvedReferences) =
        try
            let tcConfig = tcConfigP.Get()
            checker.FrameworkImportsCache.Get tcConfig
        with e ->
            stopProcessingRecovery e range0; failwithf "Error creating evaluation session: %A" e
````

I also tried to buils and test with my change of the global checker `Create`

````fsharp
    let checker = FSharpChecker.Create(msbuildEnabled = false)
````

The result/exception is the same.

Now it is getting too cryptic for me. I can make some changes in FCS, build,
and try on a "clean" test machine (I have just one). If you want me to make
some diagnostic changes and test them, please let me know.

**Comment by dsyme** [comment]
@nightroman Thanks for the update.  

On the first question: using the global ``FSharpChecker.Instance`` is deprecated - please create an FSharpChecker.  Then you can choose whether to enable msbuild (meaning enable-if-installed)  or not

On the second question.  If you are not using the ProjectCracker, and you are using ``msbuildEnabled = false``, then you should never be resolving Microsoft.MSBuild.Framework (unless you are actually referencing that in your F# compilation).    

Any chance you can list repro steps?  I'd be happy to dig into it to help land this cleanly.




Are you using the ProjectCracker?

**Comment by nightroman** [comment]
@dsyme On the first question: please note that I do not use deprecated FSharpChecker.Instance. But it is always created by the library in the **static** constructor of FSharpChecker. Thus MSBuild is always loaded if it is present, even if I opt out.

> Are you using the ProjectCracker?
No. I have the dependency but I do not call it in scenarios related to this topic. MSBuild should not be loaded in my scenarios, per my understanding of 8.0.0 and my use of msbuildEnabled = false. But it happens to be loaded.

**Comment by dsyme** [comment]
> But it is always created by the FCS library itself in the static initializer of FSharpChecker. Thus MSBuild is always loaded if it is present, even if I opt out.

Ah, I see.  On a machine without MSBuild this attempted  load should not cause a failure since it gets caught.  So some other exception is happening. Can you share the stack trace and details of the exception?

**Comment by nightroman** [comment]
It does not fail in Create(). As I wrote above, it fails in try/with here

FSharp.Compiler.Service\src\fsharp\fsi\fsi.fs:

````fsharp
    /// The single, global interactive checker that can be safely used in conjunction with other operations
    /// on the FsiEvaluationSession. 
    let checker = FSharpChecker.Create()

    let (tcGlobals,frameworkTcImports,nonFrameworkResolutions,unresolvedReferences) =
        try
            let tcConfig = tcConfigP.Get()
            checker.FrameworkImportsCache.Get tcConfig
        with e ->
            stopProcessingRecovery e range0; failwithf "Error creating evaluation session: %A" e
````

That is all I can currently get, not much, unfortunately.

**Comment by nightroman** [comment]
I will remove try/with, build, and try again on my the only "clean" machine. Hopefully, we will get the stack. I am sorry that my ways to diagnose this issue are very limited.

**Comment by dsyme** [comment]
ok thanks

**Comment by nightroman** [comment]
Here is the call stack

````
TypeInitializationException:
The type initializer for 'Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker' threw an exception.

TypeInitializationException:
The type initializer for '<StartupCode$FSharp-Compiler-Service>.$Service' threw an exception.

InvalidOperationException:
unexpected msbuildEnabled

System.TypeInitializationException: The type initializer for 'Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker' threw an exception. ---> System.TypeInitializationException: The type initializer for '<StartupCode$FSharp-Compiler-Service>.$Service' threw an exception. ---> System.InvalidOperationException: unexpected msbuildEnabled
   at Microsoft.FSharp.Compiler.SimulatedMSBuildReferenceResolver.GetBestAvailableResolver(Boolean msbuildEnabled) in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\SimulatedMSBuildReferenceResolver.fs:line 169
   at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker.Create(FSharpOption`1 projectCacheSize, FSharpOption`1 keepAssemblyContents, FSharpOption`1 keepAllBackgroundResolutions, FSharpOption`1 msbuildEnabled) in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\vs\service.fs:line 2740
   at <StartupCode$FSharp-Compiler-Service>.$Service..cctor() in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\vs\service.fs:line 2721
   --- End of inner exception stack trace ---
   at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker..cctor() in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\vs\service.fs:line 0
   --- End of inner exception stack trace ---
   at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpChecker.Create(FSharpOption`1 projectCacheSize, FSharpOption`1 keepAssemblyContents, FSharpOption`1 keepAllBackgroundResolutions, FSharpOption`1 msbuildEnabled)
   at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession..ctor(FsiEvaluationSessionHostConfig fsi, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, Boolean fsiCollectible, Boolean msbuildEnabled) in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\fsi\fsi.fs:line 2462
   at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession.Create(FsiEvaluationSessionHostConfig fsiConfig, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, FSharpOption`1 collectible, FSharpOption`1 msbuildEnabled) in C:\-\GIT\fs-proj\FSharp.Compiler.Service\src\fsharp\fsi\fsi.fs:line 2725
   at FSharpFar.Session.Session..ctor(String configFile)
   at FSharpFar.Session.Session.FindOrCreate(String path)
   at FSharpFar.FarCommand.Invoke(Object sender, ModuleCommandEventArgs e)
   at FarNet.Works.ProxyCommand.Invoke(Object sender, ModuleCommandEventArgs e)
   at FarNet.Far0.InvokeCommand(Char* command, Boolean isMacro)
   at FarNet.Far0.AsOpen(OpenInfo* info)
   at OpenW(OpenInfo* info)
````

**Comment by nightroman** [comment]
I replaced:

src/fsharp/fsi/fsi.fs

````
    let checker = FSharpChecker.Create()

    let (tcGlobals,frameworkTcImports,nonFrameworkResolutions,unresolvedReferences) = 
        try 
            let tcConfig = tcConfigP.Get()
            checker.FrameworkImportsCache.Get tcConfig
        with e -> 
            stopProcessingRecovery e range0; failwithf "Error creating evaluation session: %A" e
````

with

````
    let checker = FSharpChecker.Create(msbuildEnabled = false)

    let (tcGlobals,frameworkTcImports,nonFrameworkResolutions,unresolvedReferences) =
        //try
            let tcConfig = tcConfigP.Get()
            checker.FrameworkImportsCache.Get tcConfig
        //with e ->
        //    stopProcessingRecovery e range0; failwithf "Error creating evaluation session: %A" e
````

and

src/fsharp/vs/service.fs

````
    static let globalInstance = FSharpChecker.Create()
````

with

````
    static let globalInstance = FSharpChecker.Create(msbuildEnabled = false)
````

Then I built FCS and ran on the problem machine.

Here is the exception:

````
    StopProcessingExn:
    Exception of type 'Microsoft.FSharp.Compiler.ErrorLogger+StopProcessingExn' was thrown.

    Microsoft.FSharp.Compiler.ErrorLogger+StopProcessingExn: Exception of type 'Microsoft.FSharp.Compiler.ErrorLogger+StopProcessingExn' was thrown.
       at Microsoft.FSharp.Compiler.Interactive.Shell.ErrorLoggerThatStopsOnFirstError.ErrorSinkHelper[a](PhasedError err)
       at Microsoft.FSharp.Compiler.Interactive.Shell.ErrorLoggerThatStopsOnFirstError.ErrorSinkImpl(PhasedError err)
       at Microsoft.FSharp.Compiler.ErrorLogger.ErrorLoggerExtensions.ErrorLogger.Error[b](ErrorLogger x, Exception exn)
       at Microsoft.FSharp.Compiler.CompileOps.TcImports.BuildFrameworkTcImports(TcConfigProvider tcConfigP, FSharpList`1 frameworkDLLs, FSharpList`1 nonFrameworkDLLs)
       at Microsoft.FSharp.Compiler.FrameworkImportsCache.Get(TcConfig tcConfig)
       at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession..ctor(FsiEvaluationSessionHostConfig fsi, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, Boolean fsiCollectible, Boolean msbuildEnabled)
       at Microsoft.FSharp.Compiler.Interactive.Shell.FsiEvaluationSession.Create(FsiEvaluationSessionHostConfig fsiConfig, String[] argv, TextReader inReader, TextWriter outWriter, TextWriter errorWriter, FSharpOption`1 collectible, FSharpOption`1 msbuildEnabled)
       at FSharpFar.Session.Session..ctor(String configFile)
       at FSharpFar.Session.Session.FindOrCreate(String path)
       at FSharpFar.FarCommand.Invoke(Object sender, ModuleCommandEventArgs e)
       at FarNet.Works.ProxyCommand.Invoke(Object sender, ModuleCommandEventArgs e)
       at FarNet.Far0.InvokeCommand(Char* command, Boolean isMacro)
       at FarNet.Far0.AsOpen(OpenInfo* info)
````

One more thing. The machine does not have F# runtime installed. I provide
FSharp.Core.dll with my package and non-FCS related F# code works fine.
Is it perhaps mandatory for FCS that F# runtime is installed?

**Comment by nightroman** [comment]
So here is the best I can get, some obscure error thrown at
````
src\fsharp\CompileOps.fs:line 4598
    error(InternalError("BuildFrameworkTcImports: no successful import of "+coreLibraryResolution.resolvedPath,coreLibraryResolution.originalReference.Range))
````

Please let me know how I can change something to improve diagnostics.

I will be able to publish my changes for 8.0.0 only next week. Then I may explain the steps to install and repro. But a clean machine is needed, too, I presume.

**Comment by nightroman** [comment]
Please, wait for some new important details. I just do not have time to explain them now.

**Comment by nightroman** [comment]
Investigated. My remaining issue was not related to MSBuild, it was about
missing *FSharp.Core.optdata* and *FSharp.Core.sigdata*.

Thus, all works fine if MSBuild is not present and the issue may be closed.
My module FSharpFar with F# interactive and editor services may work in Far
Manager without installing anything but the module itself. This is great!

There is one minor imperfection though, as shown before. If MSBuild is present
then it is *loaded* even if I opt out. This is happening due to two global calls
to `FSharpChecker.Create()` with default parameters:

- [src/fsharp/fsi/fsi.fs](https://github.com/fsharp/FSharp.Compiler.Service/blob/8.0.0/src/fsharp/fsi/fsi.fs#L2462)
- [src/fsharp/vs/service.fs](https://github.com/fsharp/FSharp.Compiler.Service/blob/8.0.0/src/fsharp/vs/service.fs#L2721)

As far as I opt out, MSBuild is not used for resolution, hopefully, it just
gets loaded. Should this be filed as a separate issue perhaps? Or is it not an
issue at all?

**Comment by nightroman** [comment]
I think the issue is resolved completely in 9.0.0. I am closing it. Thank you!
